### DISCUSSION TITLE
🚀 Born v0.6.0 Released: ONNX Import & Lazy GPU Mode

### DISCUSSION THREAD

**Discussion by kolkov** [born-ml/born#12] [Announcements]
# Born v0.6.0: ONNX Import & Lazy GPU Mode

We're excited to announce **Born v0.6.0** — a major release that brings ONNX model import and GPU-resident lazy evaluation to the Born ML framework.

### 🔄 ONNX Import API

Born can now load and run models trained in PyTorch, TensorFlow, or any ONNX-compatible framework:

```go
import "github.com/born-ml/born/internal/onnx"

// Load ONNX model
model, err := onnx.LoadModel("model.onnx")
if err != nil {
    log.Fatal(err)
}

// Run inference
output, err := model.Forward(inputTensor)
```

**Features:**
- Full ONNX protobuf parser
- 30+ standard operators (Add, MatMul, Conv2D, Relu, Softmax, Reshape, Transpose, etc.)
- Extensible operator registry for custom ops
- Automatic weight loading and graph construction

### ⚡ Lazy GPU Evaluation

Tensors now stay on GPU until explicitly needed, eliminating unnecessary CPU-GPU transfers:

```go
import "github.com/born-ml/born/backend/webgpu"

backend := webgpu.New()

// All operations stay on GPU
a := tensor.Randn[float32](tensor.Shape{1024, 1024}, backend)
b := tensor.Randn[float32](tensor.Shape{1024, 1024}, backend)
c := a.MatMul(b)           // GPU computation, no transfer
d := c.Add(a).Mul(b)       // Chained ops, still on GPU

// Data transfers only when needed
result := d.Data()          // NOW transfers to CPU
```

**Performance improvements:**
- **~100x fewer GPU submits** via command batching
- **Zero CPU copies** for chained operations
- **Automatic memory management** with `runtime.SetFinalizer`

### 🔧 Bug Fixes

- Fixed GPU→CPU→GPU copy inefficiency with direct `CopyBufferToBuffer`
- Fixed memory leak — GPU buffers now properly released after `Realize()`
- Updated for golangci-lint v2.7.0 compatibility

## Installation

```bash
go get github.com/born-ml/born@v0.6.0
```

## What's Next

With ONNX import complete, we're moving toward:
- **Phase 7**: Transformer inference optimization
- **Phase 8**: Production deployment tools
- **Phase 9**: Distributed training support

## Links

- 📦 [Release Notes](https://github.com/born-ml/born/releases/tag/v0.6.0)
- 📖 [Documentation](https://github.com/born-ml/born#readme)
- 🗺️ [Roadmap](https://github.com/born-ml/born/blob/main/ROADMAP.md)

---

Thank you to everyone who contributed to this release! Questions and feedback are welcome in the discussions below.
